esx_service/vmdk_ops.py

```diff
@@ -1670,22 +1683,39 @@ def set_vol_opts(name, tenant_name, options):
 
     return False
 
+def wait_ops_in_flight():
+    global STOP_RETRY_COUNT
+
+    # Set the STOP flag to true
+    stopEvent.set()
+
+    # Wait until the in-flight ops counter goes down to zero
+    while ((opsCounter.value > 0) and (STOP_RETRY_COUNT > 0)):
+        logging.info("Service is being stopped. Waiting for completion of %d operations in flight...", opsCounter.value)
+        time.sleep(STOP_RETRY_SLEEP)
```

Comment:
wait_ops_in_flight() is a signal handler and will be executed on the main thread. 
Which means main thread will be blocked until you exit from `wait_ops_in_flight`, so I expect ALL volume-related docker commands to hang once you started the drain.

In the other words, I do not believe this could work. You should set the flag and spawn the thread for waiting, and exit on thread exit. Or you can do it fully async and get notified (semaphor or conditional) when the counter changes. 
But you should not block in main thread.

Reply:
Do you mean all the ongoing logic in the main thread will be blocked when a signal handler is executed (for us, it mean the "while True" logic in handleVmciRequests() will be blocked by this)? I didn't see this description in Python signal documentation though.

Reply:
wait_ops_in_flight() should just have to set the stop event and then return.  The loop in handleVmciRequests() should check for the stop event and exit that loop and wait for the command count to drain to zero. the signal handler shouldn't have to do anything else besides setting the stop event.

Reply:
@shaominchen  0 https://docs.python.org/2/library/signal.html says

Some care must be taken if both signals and threads are used in the same program. The fundamental thing to remember in using signals and threads simultaneously is: always perform signal() operations in the main thread of execution. Any thread can perform an alarm(), getsignal(), pause(), setitimer() or getitimer(); only the main thread can set a new signal handler, and the **main thread will be the only one to receive signals (this is enforced by the Python signal module, even if the underlying thread implementation supports sending signals to individual threads)***.

I read it that if the main thread executes the signal handler and the code stays in the handler forever, nothing else will be executed on the main thread

Reply:
I will do some more investigation/testing about this.

Reply:
@govint Please see my latest answer to @msterin 's comments below.